### PR TITLE
Nonconforming interfaces struct

### LINKED ISSUE #1133: Incorporate NCIs into assembly
## Tasks

- [ ] add NCI container, using #1177 and #1174 to populate views
- [ ] add test post-assembly-construction for interface being properly made (for now, use simple mesh (3-elem). more complex geometry can be done in a future PR

## Acceptance Criteria

- [ ] Assembly structs are populated with values needed for fluid-solid NCI
- [ ] Test passes

### LINKED ISSUE #1177: NCI normals
## Task description

Write a list of things that need to be completed to achieve the goal of this issue.

This likely depends on the `coupled_interface` rewrite.
- [ ] Appropriate `coupled_interface` normal computation into NCIs (migrate computation into external function and delegate to that in `coupled_interface`?)
- [ ] Write test to validate NCI normal / expand `coupled_interface` normal test to NCIs:
  - orthogonal to edge (interpolate + finite difference along-edge; dot product approximately zero)
  - ensure unit length
  - outward facing (interpolate + finite difference towards interior; dot product less than zero)

## Acceptance criteria
- [ ] The test passes

### LINKED ISSUE #1251: Add tests for 1D Jacobian along edges
- [ ] Add a test for edge/intersection factor that provides the correct jacobian times weight. Summing these factors should recover the length of the interface.

### PR DESCRIPTION
**Pull request by int-ptr-ptr**
## Description

Please describe the changes/features in this pull request.

## Issue Number

Closes #1133 

## Checklist

Please make sure to check developer documentation on specfem docs.

- [x] I ran the code through pre-commit to check style
- [x] **THE DOCUMENTATION BUILDS WITHOUT WARNINGS/ERRORS**
- [x] I have added labels to the PR (see right hand side of the PR page)
- [ ] My code passes all the integration tests
- [ ] I have added sufficient unittests to test my changes
- [ ] I have added/updated documentation for the changes I am proposing
- [ ] I have updated CMakeLists to ensure my code builds
- [ ] My code builds across all platforms

### CODE REVIEW

**Comment by buildbot-princeton** [comment]
Can one of the admins verify this patch?

**Comment by buildbot-princeton** [comment]
Can one of the admins verify this patch?

**Comment by buildbot-princeton** [comment]
Can one of the admins verify this patch?

**Comment by Rohit-Kakodkar** [comment]
ok to test

**Comment by int-ptr-ptr** [comment]
```
fatal error: translation unit is too large for Clang to process: ran out of source locations
PLEASE append the compiler options "-save-temps -v", rebuild the application to get the full command which is failing and submit a bug report to https://software.intel.com/en-us/support/priority-support which includes the failing command, input files for the command and the crash backtrace (if any).
```

This gets fixed by `"SPECFEM_ENABLE_UNITY_BUILD": "OFF"`. @Rohit-Kakodkar, is it okay to disable unity for LLVM?

**Comment by Rohit-Kakodkar** [comment]
> ```
> fatal error: translation unit is too large for Clang to process: ran out of source locations
> PLEASE append the compiler options "-save-temps -v", rebuild the application to get the full command which is failing and submit a bug report to https://software.intel.com/en-us/support/priority-support which includes the failing command, input files for the command and the crash backtrace (if any).
> ```
> 
> This gets fixed by `"SPECFEM_ENABLE_UNITY_BUILD": "OFF"`. @Rohit-Kakodkar, is it okay to disable unity for LLVM?

See https://github.com/PrincetonUniversity/SPECFEMPP/blob/3e33f18273a246834a83a4fd0a4d5d56ddf9eb29/CMakeLists.txt#L423

You can specify the size of translational units based on compiler type.

File: core/specfem/assembly/coupled_interfaces/dim2/coupled_interface.cpp [hunk 1]
```diff
@@ -26,6 +28,17 @@ specfem::assembly::coupled_interfaces<specfem::dimension::type::dim2>::
...
+                       INTERFACE_TAG(ELASTIC_ACOUSTIC, ACOUSTIC_ELASTIC),
+                       BOUNDARY_TAG(NONE, STACEY, ACOUSTIC_FREE_SURFACE,
+                                    COMPOSITE_STACEY_DIRICHLET)),
+                      CAPTURE(interface_container) {
+                        _interface_container_ =
+                            NonconformingInterfaceContainerType<_interface_tag_,
+                                                                _boundary_tag_>(
+                                ngllz, ngllx, edge_types, jacobian_matrix,
+                                mesh);
+                      })
```
Comment (lsawade) [inline core/specfem/assembly/coupled_interfaces/dim2/coupled_interface.cpp]: Why can this not be combined? the interfacecontainertype is templated by the interface_tag, no?

File: core/specfem/assembly/coupled_interfaces/dim2/coupled_interface.hpp [hunk 2]
```diff
@@ -62,6 +69,13 @@ class coupled_interfaces<specfem::dimension::type::dim2>
                       DECLARE(((InterfaceContainerType,
                                 (_INTERFACE_TAG_, _BOUNDARY_TAG_)),
                                interface_container)))
+  FOR_EACH_IN_PRODUCT((DIMENSION_TAG(DIM2), CONNECTION_TAG(NONCONFORMING),
+                       INTERFACE_TAG(ELASTIC_ACOUSTIC, ACOUSTIC_ELASTIC),
+                       BOUNDARY_TAG(NONE, ACOUSTIC_FREE_SURFACE, STACEY,
+                                    COMPOSITE_STACEY_DIRICHLET)),
+                      DECLARE(((NonconformingInterfaceContainerType,
+                                (_INTERFACE_TAG_, _BOUNDARY_TAG_)),
+                               interface_container)))
```
Comment (lsawade) [inline core/specfem/assembly/coupled_interfaces/dim2/coupled_interface.hpp]: See other comments

File: core/specfem/assembly/nonconforming_interfaces/dim2/impl/compute_intersection.tpp [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        specfem::point::global_coordinates<specfem::dimension::type::dim2> *,
+        Kokkos::HostSpace> &element2,
+    const specfem::mesh_entity::type &edge1,
+    const specfem::mesh_entity::type &edge2,
+    const Kokkos::View<type_real *, Kokkos::HostSpace> &mortar_quadrature,
+    const Kokkos::View<type_real *, Kokkos::HostSpace> &element_quadrature,
+    TransferView1 &transfer_function1, TransferView2 &transfer_function2) {
+  const auto nquad_mortar = mortar_quadrature.extent(0);
+  const auto ngll = element_quadrature.extent(0);
+  Kokkos::View<type_real **, Kokkos::HostSpace> devnull("prime_capture",
```
Comment (Rohit-Kakodkar) [inline core/specfem/assembly/nonconforming_interfaces/dim2/impl/compute_intersection.tpp:117]: Can you create 2 `devnulls` here to pass to the next function. Ideally, it should not matter, but the compiler could miss some optimizations since accesses within the next function refer to the same memory address. However, ideally compiler to optimize out the allocations since they are discarded on exit from the function.

File: tests/unit-tests/assembly/nonconforming_interfaces/container_init_test.cpp [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+      << edgecoord << ".";
+}
+
+void test_nonconforming_container_transfers(
+    const specfem::assembly::assembly<specfem::dimension::type::dim2>
+        &assembly) {
+  const type_real length_verify_epsilon = 1e-2;
+
+  const int ngllx = assembly.mesh.shape_functions::ngllx;
+  const int ngllz = assembly.mesh.shape_functions::ngllz;
```
Comment (Rohit-Kakodkar) [inline tests/unit-tests/assembly/nonconforming_interfaces/container_init_test.cpp]:
```suggestion
  const int ngllx = assembly.mesh.element_grid.ngllx;
  const int ngllz = assembly.mesh.element_grid.ngllz;
```

**Review by Rohit-Kakodkar — APPROVED** [review]
Minor changes, apart from the thing that Lucas mentioned - make non-conforming container as a specialization for a general `interface_container`

File: core/specfem/assembly/CMakeLists.txt [hunk 5]
```diff
@@ -20,6 +20,13 @@ target_link_libraries(
         ${BOOST_LIBS}
 )
 
+if (__INTEL_LLVM__)
+    set_target_properties(assembly PROPERTIES
+            # UNITY_BUILD OFF
+            UNITY_BUILD_BATCH_SIZE 2
+    )
+endif()
```
Comment (lsawade) [inline core/specfem/assembly/CMakeLists.txt]:
```suggestion
set_target_properties(assembly PROPERTIES
    # UNITY_BUILD $<NOT:$<BOOL:${__INTEL_LLVM__}>>
    UNITY_BUILD_BATCH_SIZE $<IF:$<BOOL:${__INTEL_LLVM__}>,2,8>
)
```

Syntax is `$<IF:condition,true-value,false-value>` and `$<BOOL:string>` is explained here: https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:BOOL

**Review by lsawade — CHANGES_REQUESTED** [review]
I don't think there is a lot of file changes involved in updating the templated interface container as discussed on the call. I think it would be better to do this here and forget about it, instead of putting it on the backburner.

Other than that minor comments only.

File: core/specfem/assembly/coupled_interfaces/dim2/coupled_interface.cpp (hunk 1 above)
Comment (int-ptr-ptr) [reply core/specfem/assembly/coupled_interfaces/dim2/coupled_interface.cpp]: `InterfaceContainerType` was not templated by `CONNECTION_TAG`, no.

I was worried about having to dive into kernels to cascade the added tag, but it ended up only being through `get_interface_container` and `load_on_device`.

**Review by lsawade — APPROVED** [review]
Thank you for making the updates!! LGTM

**Review by icui — APPROVED** [review]
